beacon-chain/blockchain/process_block.go

```diff
@@ -432,21 +439,31 @@ func (s *Service) onBlockBatch(ctx context.Context, blks []interfaces.SignedBeac
 				return err
 			}
 		}
-
-		if err := s.insertBlockToForkChoiceStore(ctx, b.Block(), blockRoots[i], fCheckpoints[i], jCheckpoints[i]); err != nil {
-			return err
-		}
-		if isValidPayload {
-			if err := s.cfg.ForkChoiceStore.SetOptimisticToValid(ctx, blockRoots[i]); err != nil {
-				return errors.Wrap(err, "could not set optimistic block to valid")
-			}
-		}
+		args := &forkchoicetypes.BlockAndCheckpoints{Block: b.Block(),
+			JustifiedEpoch: jCheckpoints[i].Epoch,
+			FinalizedEpoch: fCheckpoints[i].Epoch}
+		pendingNodes[len(blks)-i-1] = args
 		s.saveInitSyncBlock(blockRoots[i], b)
 		if err = s.handleBlockAfterBatchVerify(ctx, b, blockRoots[i], fCheckpoints[i], jCheckpoints[i]); err != nil {
 			tracing.AnnotateError(span, err)
 			return err
 		}
 	}
+	// Insert all nodes but the last one to forkchoice
```

Comment:
It looks like you are inserting all the nodes here. What am I missing?

Reply:
InsertOptimisticChain does not insert the last block

Reply:
Interesting. I'd think it should be the caller's responsibility to not insert the last block and the method itself inserts everything. Hopefully someone doesn't mess this up later!

Reply:
The problem is that we can't have all block roots. You get the htr of one block as the parentRoot of the next one. So the last block that doesn't have a child has an unknown root.